src/schema/sale_artwork.ts

```diff
@@ -364,4 +382,24 @@ const SaleArtwork: GraphQLFieldConfig<void, ResolverContext> = {
   },
 }
 
+const shouldFetchArtwork = (
+  fieldNode: Readonly<FieldNode>,
+  fragments: any
+): boolean => {
+  const fetchArtworkFields = ["edition_of", "edition_sets"]
+  const requestedArtworkFields: string[] = []
+  visit(fieldNode, {
+    Field(node) {
+      requestedArtworkFields.push(node.name.value)
+    },
+    FragmentSpread(node) {
+      const fragmentDef = fragments[node.name.value]
+      requestedArtworkFields.push(
+        ...fragmentDef.selectionSet.selections.map(s => s.name.value)
+      )
+    },
+  })
+  return requestedArtworkFields.some(f => fetchArtworkFields.includes(f))
+}
```

Comment:
looks like `visit` kind of visit's the fields recursively, do you know if there is a way to limit it to only one level deep?

Reply:
@alloy updated a logic a little should be slightly better since i don't check if we already know we need to fetch, but i think `visit` still traverses all nodes.

Reply:
Check the source (and tests) for the best docs https://github.com/graphql/graphql-js/blob/61f993604bbc3e0aa2dcfc9191b6f6d4cb060bee/src/language/visitor.js#L135-L165

So you should be able to return `false` to stop the visitor from recursing further down into a sub-selection and you can probably do that based on the `path` parameter to the visitor function.

Additionally, I think you should make your visitor recursive when it encounters a fragment, as a fragment could be spread into a fragment into a fragment and so on… 🐢 

```js
	const visitor = (rootNode: FieldNode) => {
	  visit(rootNode, {
	    Field(node, _key, _parent, path, _ancestors) {
	      if (path.length > 1) {
	        return false
	      }
	      requestedArtworkFields.push(node.name.value)
	    },
	    FragmentSpread(node) {
	      const fragmentDef = fragments[node.name.value]
	      // Recurse into fragment
	      visitor(fragmentDef)
	    },
	  })
	}
```

(This is untested code!)

Reply:
@alloy let me know what you think about this approach? had to check `path.length <= 3` since for fragments we have `selectionSets.selections` also in the path.

Reply:
You’re not actually visiting the fragment yet, though. See the second comment in my example code above or let me know if it still isn’t clear 🙏

Reply:
oh meant to ask you about this, right now im checking fragment's field manually within that block so this kind of takes care of first level fields. But i guess you are talking about 
```graphql
fragment TestF1 on Artwork {
  edition_of
}
fragment TestF2 on Artwork {
 id
}
query {
  artwork(id: "...") {
     ...testF1 {
       ....testF2 
     }
  }
}
```
is this even valid?

Reply:
Sure is!

Reply:
Well, actually TestF2 would spread inside TestF1.

Reply:
trying this now, i don't think we can call original `visit` method here, we basically need to right the visit method or use `shouldFetchArtwork`
```javascript
FragmentSpread(node) {
      const fragmentDef = fragments[node.name.value]
      visit(fragmentDef, {
        Field(node) {
          ...
        }
      })
    },
```
or 
```js
 FragmentSpread(node) {
      const fragmentDef = fragments[node.name.value]
      needsArtworkFetch = shouldFetchArtwork(fragmentDef, fragments)
    },
```

Reply:
Hmm, not sure I follow, if you would define the function inline then the fragments should still be in scope, right?

Reply:
Btw, we could probably extract this code afterwards and use it for other places where we’ve been accessing things manually (eg in the blacklist fields function).

Reply:
yes, @sweir27 mentioned the same thing about extracting it, let me see if i find a good common place for it.